### ISSUE TITLE
LOGICAL_ERROR: Bad cast from type DB::ColumnVector<char8_t> to DB::ColumnVector<signed char>

### ISSUE THREAD

**Issue by JZuming** [ClickHouse/ClickHouse#50284]
**Describe what's wrong**

The SELECT statement should output results normally, but it threw an exception.

**Does it reproduce on recent release?**

It can be reproduced in the latest version.

**How to reproduce**

Version: 23.5.1.1 (commit 3e6314675c6467bc4dd78f659bac862f7e9648f8)

Easy reproduce in ClickHouse fiddle: https://fiddle.clickhouse.com/4996cfc2-e407-4c7e-b9e9-de67e040bbc9

_Set up database_

```sql
create table t0 (pkey UInt32, c1 UInt32, primary key(pkey)) engine = MergeTree;
create table t1 (vkey UInt32, primary key(vkey)) engine = MergeTree;
create table t3 (c17 String, primary key(c17)) engine = MergeTree;
insert into t1 values (3);
```

_Bug-triggering query_

```sql
WITH 
cte_1 AS (select  
    subq_1.c_5_c1698_16 as c_2_c1702_3, 
    subq_1.c_5_c1694_12 as c_2_c1703_4
  from 
    (select  
          covarPop(-0, 74) as c_5_c1686_4, 
          sumWithOverflow(0) as c_5_c1694_12, 
          covarPop(-53.64, 92.63) as c_5_c1698_16
        from 
          t3 as ref_8
        group by ref_8.c17) as subq_1)
select 
    ref_15.c_2_c1703_4 as c_2_c1723_6, 
    ref_15.c_2_c1702_3 as c_2_c1724_7
  from 
    t0 as ref_14
        RIGHT outer join cte_1 as ref_15
        on (ref_14.c1 = ref_15.c_2_c1702_3)
      RIGHT outer join t1 as ref_16
      on (ref_14.pkey = ref_16.vkey);
```

**Expected behavior**

The SELECT statement is executed without any error.

**Actual behavior**

It throws an exception:
```
Received exception from server (version 23.4.2):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: There's no column 'subq_1.c_5_c1698_16' in table 'subq_1': While processing subq_1.c_5_c1698_16 AS c_2_c1702_3. (UNKNOWN_IDENTIFIER)
(query: WITH 
cte_1 AS (select  
    subq_1.c_5_c1698_16 as c_2_c1702_3, 
    subq_1.c_5_c1694_12 as c_2_c1703_4
  from 
    (select  
          covarPop(-0, 74) as c_5_c1686_4, 
          sumWithOverflow(0) as c_2_c1703_4, 
          covarPop(-53.64, 92.63) as c_2_c1702_3
        from 
          t3 as ref_8
        group by ref_8.c17) as subq_1)
select 
    ref_15.c_2_c1703_4 as c_2_c1723_6, 
    ref_15.c_2_c1702_3 as c_2_c1724_7
  from 
    t0 as ref_14
        RIGHT outer join cte_1 as ref_15
        on (ref_14.c1 = ref_15.c_2_c1702_3)
      RIGHT outer join t1 as ref_16
      on (ref_14.pkey = ref_16.vkey);)
```

**Additional context**

The earliest reproducible version is 21 in fiddle: https://fiddle.clickhouse.com/cdaec019-a5f5-4c1b-8424-bf8c5662b435

**Comment by Avogar** [comment]
This query works with enabled new analyzer:
```
avogar-dev :) set allow_experimental_analyzer=1

SET allow_experimental_analyzer = 1

Query id: e42f9e89-c1df-484a-8d69-358c78f2cd60

Ok.

0 rows in set. Elapsed: 0.001 sec.

avogar-dev :) WITH
cte_1 AS (select
    subq_1.c_5_c1698_16 as c_2_c1702_3,
    subq_1.c_5_c1694_12 as c_2_c1703_4
  from
    (select
          covarPop(-0, 74) as c_5_c1686_4,
          sumWithOverflow(0) as c_5_c1694_12,
          covarPop(-53.64, 92.63) as c_5_c1698_16
        from
          t3 as ref_8
        group by ref_8.c17) as subq_1)
select
    ref_15.c_2_c1703_4 as c_2_c1723_6,
    ref_15.c_2_c1702_3 as c_2_c1724_7
  from
    t0 as ref_14
        RIGHT outer join cte_1 as ref_15
        on (ref_14.c1 = ref_15.c_2_c1702_3)
      RIGHT outer join t1 as ref_16
      on (ref_14.pkey = ref_16.vkey);

┌─c_2_c1723_6─┬─c_2_c1724_7─┐
│           0 │           0 │
└─────────────┴─────────────┘

1 row in set. Elapsed: 0.004 sec.
```

Most likely it's a bug in old infrastructure with query analisis. Let's look at the result of explain:
```

┌─explain─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Expression (( + (Projection + Before ORDER BY)))                                                                                                                    │
│ Header: ref_15.c_2_c1703_4 Int8                                                                                                                                     │
│         ref_15.c_2_c1702_3 Float64                                                                                                                                  │
│ Actions: INPUT : 0 -> c_2_c1702_3 Float64 : 0                                                                                                                       │
│          INPUT : 1 -> c_2_c1703_4 Int8 : 1                                                                                                                          │
│          ALIAS c_2_c1702_3 :: 0 -> ref_15.c_2_c1702_3 Float64 : 2                                                                                                   │
│          ALIAS c_2_c1703_4 :: 1 -> ref_15.c_2_c1703_4 Int8 : 0                                                                                                      │
│ Positions: 0 2                                                                                                                                                      │
│   Join (JOIN FillRightFirst)                                                                                                                                        │
│   Header: pkey UInt32                                                                                                                                               │
│           c_2_c1702_3 Float64                                                                                                                                       │
│           c_2_c1703_4 Int8                                                                                                                                          │
│     Expression ((Before JOIN + ( + (Projection + Before ORDER BY))))                                                                                                │
│     Header: pkey UInt32                                                                                                                                             │
│             c_2_c1702_3 Float64                                                                                                                                     │
│             c_2_c1703_4 Int8                                                                                                                                        │
│     Actions: INPUT :: 0 -> pkey UInt32 : 0                                                                                                                          │
│              INPUT :: 1 -> c_2_c1702_3 Float64 : 1                                                                                                                  │
│              INPUT :: 2 -> c_2_c1703_4 Int8 : 2                                                                                                                     │
│     Positions: 0 1 2                                                                                                                                                │
│       Join (JOIN FillRightFirst)                                                                                                                                    │
│       Header: pkey UInt32                                                                                                                                           │
│               _CAST(c1, Float64) Float64                                                                                                                            │
│               c_2_c1703_4 Int8                                                                                                                                      │
│               c_2_c1702_3 Float64                                                                                                                                   │
│         Expression ((Convert JOIN columns + Before JOIN))                                                                                                           │
│         Header: pkey UInt32                                                                                                                                         │
│                 _CAST(c1, Float64) Float64                                                                                                                          │
│         Actions: INPUT :: 0 -> pkey UInt32 : 0                                                                                                                      │
│                  INPUT : 1 -> c1 UInt32 : 1                                                                                                                         │
│                  COLUMN Const(String) -> Float64 String : 2                                                                                                         │
│                  FUNCTION _CAST(c1 :: 1, Float64 :: 2) -> _CAST(c1, Float64) Float64 : 3                                                                            │
│         Positions: 0 3                                                                                                                                              │
│           ReadFromPreparedSource (Read from NullSource)                                                                                                             │
│           Header: pkey UInt32                                                                                                                                       │
│                   c1 UInt32                                                                                                                                         │
│         Expression ((Convert joined columns + (Joined actions + (Rename joined columns + ( + (Projection + (Before ORDER BY + (Projection + Before ORDER BY)))))))) │
│         Header: c_2_c1702_3 Float64                                                                                                                                 │
│                 c_2_c1703_4 Int8                                                                                                                                    │
│         Actions: INPUT : 0 -> sumWithOverflow(0) UInt8 : 0                                                                                                          │
│                  INPUT : 1 -> covarPop(-53.64, 92.63) Float64 : 1                                                                                                   │
│                  ALIAS sumWithOverflow(0) :: 0 -> c_5_c1694_12 UInt8 : 2                                                                                            │
│                  ALIAS covarPop(-53.64, 92.63) :: 1 -> c_5_c1698_16 Float64 : 0                                                                                     │
│                  ALIAS c_5_c1694_12 :: 2 -> c_2_c1703_4 Int8 : 1                                                                                                    │
│                  ALIAS c_5_c1698_16 :: 0 -> c_2_c1702_3 Float64 : 2                                                                                                 │
│         Positions: 2 1                                                                                                                                              │
│           Aggregating                                                                                                                                               │
│           Header: c17 String                                                                                                                                        │
│                   sumWithOverflow(0) UInt8                                                                                                                          │
│                   covarPop(-53.64, 92.63) Float64                                                                                                                   │
│           Keys: c17                                                                                                                                                 │
│           Aggregates:                                                                                                                                               │
│               sumWithOverflow(0)                                                                                                                                    │
│                 Function: sumWithOverflow(UInt8) → UInt8                                                                                                            │
│                 Arguments: 0                                                                                                                                        │
│               covarPop(-53.64, 92.63)                                                                                                                               │
│                 Function: covarPop(Float64, Float64) → Float64                                                                                                      │
│                 Arguments: -53.64, 92.63                                                                                                                            │
│           Skip merging: 0                                                                                                                                           │
│             Expression (Before GROUP BY)                                                                                                                            │
│             Header: c17 String                                                                                                                                      │
│                     0 UInt8                                                                                                                                         │
│                     -53.64 Float64                                                                                                                                  │
│                     92.63 Float64                                                                                                                                   │
│             Actions: INPUT :: 0 -> c17 String : 0                                                                                                                   │
│                      COLUMN Const(UInt8) -> 0 UInt8 : 1                                                                                                             │
│                      COLUMN Const(Float64) -> -53.64 Float64 : 2                                                                                                    │
│                      COLUMN Const(Float64) -> 92.63 Float64 : 3                                                                                                     │
│             Positions: 0 1 2 3                                                                                                                                      │
│               ReadFromPreparedSource (Read from NullSource)                                                                                                         │
│               Header: c17 String                                                                                                                                    │
│     Expression ((Joined actions + (Rename joined columns + (Projection + Before ORDER BY))))                                                                        │
│     Header: vkey UInt32                                                                                                                                             │
│     Actions: INPUT :: 0 -> vkey UInt32 : 0                                                                                                                          │
│     Positions: 0                                                                                                                                                    │
│       ReadFromMergeTree (default.t1)                                                                                                                                │
│       Header: vkey UInt32                                                                                                                                           │
│       ReadType: Default                                                                                                                                             │
│       Parts: 1                                                                                                                                                      │
│       Granules: 1                                                                                                                                                   │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Interesting part:
```
│         Actions: INPUT : 0 -> sumWithOverflow(0) UInt8 : 0                                                                                                          │
│                  INPUT : 1 -> covarPop(-53.64, 92.63) Float64 : 1                                                                                                   │
│                  ALIAS sumWithOverflow(0) :: 0 -> c_5_c1694_12 UInt8 : 2                                                                                            │
│                  ALIAS covarPop(-53.64, 92.63) :: 1 -> c_5_c1698_16 Float64 : 0                                                                                   │
│                  ALIAS c_5_c1694_12 :: 2 -> c_2_c1703_4 Int8 : 1                                                                                                    │
│                  ALIAS c_5_c1698_16 :: 0 -> c_2_c1702_3 Float64 : 2
```

For some reason, alias `c_2_c1703_4` of `c_5_c1694_12` changed type from `UInt8` to `Int8`. And it's difficult to say why it happend. Maybe @KochetovNicolai can know more details

**Comment by JZuming** [comment]
@alexey-milovidov This bug is also fixed by #50536
